python_spike/caseprocessor/decorators/transaction_decorators.py

```diff
@@ -0,0 +1,51 @@
+from functools import wraps
+from typing import Callable
+import logging
+from db.db_utility import Session
+
+
+'''
+Example callback function signature:
+
+@pubsub_transaction
+@sqlalchemy_transaction
+def callback(message: pubsub_v1.subscriber.message.Message, session: Session) -> None:
+    ... 
+
+'''
+ 
+def pubsub_transaction(sql_transaction: Callable) -> Callable:
+    @wraps(sql_transaction)
+    def with_pubsub_error_handling(*args, **kwargs):
+        pubsub_message = args[0]
+
+        try:
+            sql_transaction(*args, **kwargs)
+            pubsub_message.ack()
+            logging.info(f"Message acked: {pubsub_message.data}")
+
+        except Exception as e:
+            logging.error(f"PubSub decorator caugth error: {e}")
+            pubsub_message.nack()
+            logging.info(f"Message nacked: {pubsub_message.data}\n")
```

Comment:
I think this needs to re-raise the exception here?

Reply:
If we raise the exception here, there's currently no way for us to handle it, which would cause a crash. Instead we want to let PubSub nack the message, hoping that next time it arrives we'll be able to process it.

Are you thinking about the case when there's a prolonged problem and we actually want it to crash?

Reply:
Ah, I see I misunderstood.

With the way this is set up, it should probably be combined into one decorator to utilise it, since it would break if you used them in the wrong order or tried to use one on its own? Perhaps that can be a future enhancement instead.